specreduce/extract.py

```diff
@@ -3,216 +3,162 @@
-        plt.imshow(self.last_img, origin='lower', aspect='auto', cmap=plt.cm.Greys_r)
-        plt.clim(np.percentile(self.last_img, (5, 98)))
-
-        plt.plot(np.arange(len(trace_line)), trace_line, c='C0')
-        plt.fill_between(
-            np.arange(len(trace_line)),
-            trace_line + self.apwidth,
-            trace_line - self.apwidth,
-            color='C0',
-            alpha=0.5
-        )
-        plt.fill_between(
-            np.arange(len(trace_line)),
-            trace_line + self.apwidth + self.skysep,
-            trace_line + self.apwidth + self.skysep + self.skywidth,
-            color='C1',
-            alpha=0.5
-        )
-        plt.fill_between(
-            np.arange(len(trace_line)),
-            trace_line - self.apwidth - self.skysep,
-            trace_line - self.apwidth - self.skysep - self.skywidth,
-            color='C1',
-            alpha=0.5
-        )
-        plt.ylim(
-            np.min(
-                trace_line - (self.apwidth + self.skysep + self.skywidth) * 2
-            ),
-            np.max(
-                trace_line + (self.apwidth + self.skysep + self.skywidth) * 2
-            )
-        )
-
-        return fig
+    # TODO: what is a reasonable default?
+    # TODO: int or float?
+    width: int = 5
+
+    # TODO: should disp_axis and crossdisp_axis be defined in the Trace object?
+
+    def __call__(self, image, trace_object, disp_axis=1, crossdisp_axis=(0,)):
+        """
+        Extract the 1D spectrum using the boxcar method.
+
+        Parameters
+        ----------
+        image : nddata-compatible image
+            image with 2-D spectral image data
+        trace_object : Trace
+            object with the trace
+        disp_axis : int
+            dispersion axis
+        crossdisp_axis : tuple (to support both 2D and 3D data)
+            cross-dispersion axis
+
+
+        Returns
+        -------
+        spec : `~specutils.Spectrum1D`
+            The extracted 1d spectrum expressed in DN and pixel units
+        """
+        # this check only applies to FlatTrace instances
+        if hasattr(trace_object, 'trace_pos'):
+            self.center = trace_object.trace_pos
+            for attr in ['center', 'width']:
+                if getattr(self, attr) < 1:
+                    raise ValueError(f'{attr} must be >= 1')
+
+        # images to use for extraction
+        wimage = _ap_weight_image(
+            trace_object,
+            self.width,
+            disp_axis,
+            crossdisp_axis,
+            image.shape)
+
+        # extract. Note that, for a cube, this is arbitrarily picking one of the
+        # spatial axis to collapse. This should be handled by the API somehow.
+        ext1d = np.sum(image * wimage, axis=crossdisp_axis)
+
+        # TODO: add uncertainty and mask to spectrum1D object
+        spec = Spectrum1D(spectral_axis=np.arange(len(ext1d)) * u.pixel,
+                          flux=ext1d * getattr(image, 'unit', u.DN))
```

Comment:
the flux unit here should inherit from the input image, if possible, and only use `u.DN` as a last case fallback. gain corrections and similar calibrations should be applied before this step. in other use cases, there may already be a flux calibration applied to 2D data or the flux may be a count rate.

Reply:
I added code to support flux units propagation from the image into the extracted spectrum.